src/net/server/middleware/tsig.rs

```diff
@@ -0,0 +1,506 @@
+            TruncationContext::HaveSigner(signer) => {
+                if let Err(err) =
+                    signer.answer(&mut additional, Time48::now())
+                {
+                    error!("Unable to sign truncated TSIG response: {err}");
+                    Err(ServiceError::InternalError)
+                } else {
+                    Ok(additional)
+                }
+            }
+
+            TruncationContext::NoSignerOnlyTheKey(key) => {
+                // We can't use the TSIG signer state we just had as that was
+                // consumed in the failed attempt to sign the answer, so we
+                // have to create a new TSIG state in order to sign the
+                // truncated response.
+                let octets = request.message().as_slice().to_vec();
+                let mut mut_msg = Message::from_octets(octets).unwrap();
+
+                match ServerTransaction::request(
+                    &key,
+                    &mut mut_msg,
+                    Time48::now(),
+                ) {
+                    Ok(None) => {
+                        error!("Unable to create signer for truncated TSIG response: internal error: request is not signed but was expected to be");
+                        Err(ServiceError::InternalError)
+                    }
+
+                    Err(err) => {
+                        error!("Unable to create signer for truncated TSIG response: {err}");
+                        Err(ServiceError::InternalError)
+                    }
+
+                    Ok(Some(signer)) => {
+                        if let Err(err) =
+                            signer.answer(&mut additional, Time48::now())
+                        {
+                            error!("Unable to sign truncated TSIG response: {err}");
+                            Err(ServiceError::InternalError)
+                        } else {
+                            Ok(additional)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn map_stream_item(
+        request: Request<RequestOctets, ()>,
+        stream_item: ServiceResult<NextSvc::Target>,
+        pp_config: &mut PostprocessingConfig<KS::Key>,
+    ) -> ServiceResult<NextSvc::Target> {
+        if let Ok(mut call_res) = stream_item {
+            if matches!(
+                call_res.feedback(),
+                Some(ServiceFeedback::BeginTransaction)
+            ) {
+                // Does it need converting from the variant that supports
+                // single messages only (ServerTransaction) to the variant
+                // that supports signing multiple messages (ServerSequence)?
+                // Note: confusingly BeginTransaction and ServerTransaction
+                // use the term "transaction" to mean completely the opppsite
+                // of each other. With BeginTransaction we mean that the
+                // caller should instead a sequence of response messages
```

Comment:
There's a typo here (not sure what it should be)

Reply:
s/opppsite/opposite/ - I'll fix it, thanks!